common/src/main/java/io/netty/util/concurrent/AbstractSchedulingEventExecutor.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright 2015 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.util.concurrent;
+
+import java.util.PriorityQueue;
+import java.util.Queue;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Abstract base class for {@link EventExecutor}s that want to support scheduling.
+ */
+public abstract class AbstractSchedulingEventExecutor extends AbstractEventExecutor {
+
+    final Queue<ScheduledFutureTask<?>> delayedTaskQueue = new PriorityQueue<ScheduledFutureTask<?>>();
```

Comment:
@trustin Maybe we even could do something like adding a protected method like:

```
protected abstract Queue<Runnable> scheduledTaskQueue();
```

This would allow the implementations to either use something like a volatile or not. For example with Embedded we not need a volatile and with SingleThreadEventExecutor we not want to do lazy at all and so safe the volatile read. WDYT ?